crates/wast/src/wast.rs

```diff
@@ -226,202 +279,169 @@ impl WastContext {
                     call,
                     message,
                 } => match self.perform_invoke(call).with_context(|| context(span))? {
-                    ActionOutcome::Returned { values } => {
+                    Outcome::Ok(values) => {
                         bail!("{}\nexpected trap, got {:?}", context(span), values)
                     }
-                    ActionOutcome::Trapped {
-                        message: trap_message,
-                    } => {
-                        if trap_message.contains(message) {
+                    Outcome::Trap(t) => {
+                        let t = t.borrow();
+                        if t.message().contains(message) {
                             continue;
                         }
                         bail!(
                             "{}\nexpected exhaustion with {}, got {}",
                             context(span),
                             message,
-                            trap_message
+                            t.message(),
                         )
                     }
                 },
                 AssertReturnCanonicalNan { span, invoke } => {
                     match self.perform_invoke(invoke).with_context(|| context(span))? {
-                        ActionOutcome::Returned { values } => {
+                        Outcome::Ok(values) => {
                             for v in values.iter() {
                                 match v {
-                                    RuntimeValue::I32(_) | RuntimeValue::I64(_) => {
-                                        bail!("{}\nunexpected integer in NaN test", context(span))
-                                    }
-                                    RuntimeValue::V128(_) => {
-                                        bail!("{}\nunexpected vector in NaN test", context(span))
-                                    }
-                                    RuntimeValue::F32(x) => {
+                                    Val::F32(x) => {
                                         if !is_canonical_f32_nan(x) {
                                             bail!("{}\nexpected canonical NaN", context(span))
                                         }
                                     }
-                                    RuntimeValue::F64(x) => {
+                                    Val::F64(x) => {
                                         if !is_canonical_f64_nan(x) {
                                             bail!("{}\nexpected canonical NaN", context(span))
                                         }
                                     }
+                                    other => panic!("expected float, got {:?}", other),
```

Comment:
Here and elsewhere in this file, we shouldn't panic on type mismatches. They indicate a bug in the test, rather than the engine being tested, but when someone is writing new tests or modifying existing ones, a plain failure is friendlier than aborting if they get something wrong.